rules.go

```diff
@@ -0,0 +1,482 @@
+		return nil, ErrSkipHTML
+	}
+
+	rule.isException = strings.HasPrefix(rule.ruleText, "@@")
+	if rule.isException {
+		rule.ruleText = rule.ruleText[2:]
+	}
+
+	if strings.Contains(rule.ruleText, "$") {
+		parts := strings.SplitN(rule.ruleText, "$", 2)
+		length := len(parts)
+
+		if length > 0 {
+			rule.ruleText = parts[0]
+		}
+
+		if length > 1 {
+			for _, option := range strings.Split(parts[1], ",") {
+				if strings.HasPrefix(option, "domain=") {
+					rule.domains = parseDomainOption(option)
+				} else {
+					optionName := strings.TrimPrefix(option, "~")
+					if ok := strings.Contains(supportedOptionsPat, optionName); !ok {
+						return nil, ErrUnsupportedRule
+					}
+					rule.options[optionName] = !strings.HasPrefix(option, "~")
+				}
+			}
+		}
+	}
+
+	rule.regexString = ruleToRegexp(rule.ruleText)
+
+	return rule, nil
+}
+
+type RuleSet struct {
+	whitelist      []*ruleAdBlock
+	whitelistRegex *regexp.Regexp
+	blacklist      []*ruleAdBlock
+	blacklistRegex *regexp.Regexp
+
+	whitelistDomainsNoOptions    []*ruleAdBlock
+	whitelistDomainsWithOptions  []*ruleAdBlock
+	whitelistIncludeDomains      map[string][]*ruleAdBlock
+	whitelistIncludeDomainsRegex map[string]*regexp.Regexp
+	whitelistExcludeDomains      map[string][]*ruleAdBlock
+	whitelistExcludeDomainsRegex map[string]*regexp.Regexp
+	blacklistDomainsNoOptions    []*ruleAdBlock
+	blacklistDomainsWithOptions  []*ruleAdBlock
+	blacklistIncludeDomains      map[string][]*ruleAdBlock
+	blacklistIncludeDomainsRegex map[string]*regexp.Regexp
+	blacklistExcludeDomains      map[string][]*ruleAdBlock
+	blacklistExcludeDomainsRegex map[string]*regexp.Regexp
+
+	whitelistIncludeOptions      map[string][]*ruleAdBlock
+	whitelistIncludeOptionsRegex map[string]*regexp.Regexp
+	whitelistExcludeOptions      map[string][]*ruleAdBlock
+	whitelistExcludeOptionsRegex map[string]*regexp.Regexp
+	blacklistIncludeOptions      map[string][]*ruleAdBlock
+	blacklistIncludeOptionsRegex map[string]*regexp.Regexp
+	blacklistExcludeOptions      map[string][]*ruleAdBlock
+	blacklistExcludeOptionsRegex map[string]*regexp.Regexp
+}
+
+func matchWhite(ruleSet RuleSet, req Request) bool {
+	if ruleSet.whitelistRegex != nil && ruleSet.whitelistRegex.MatchString(req.URL.String()) {
+		return true
+	}
+
+	options := extractOptionsFromRequest(req)
+	for option, active := range options {
+		if ruleSet.whitelistIncludeOptionsRegex[option] != nil && ruleSet.whitelistIncludeOptionsRegex[option].MatchString(req.URL.String()) {
+			return active == true
+		}
+		if ruleSet.whitelistExcludeOptionsRegex[option] != nil && ruleSet.whitelistExcludeOptionsRegex[option].MatchString(req.URL.String()) {
+			return active == false
+		}
+	}
+	return false
+}
+
+func matchBlack(ruleSet RuleSet, req Request) bool {
+	if ruleSet.blacklistRegex != nil && ruleSet.blacklistRegex.MatchString(req.URL.String()) {
+		return true
+	}
+
+	rulesIncluded := []*ruleAdBlock{}
+	for _, rule := range ruleSet.blacklistDomainsNoOptions {
```

Comment:
Am I correct that you loop through a list of all rules without options, take these rules, compile new regexp and compare?

Sorry, Denis, this is not acceptable. This is really misleading and have the same performance as simple looping through the list of all rules and matching them.

Reply:
Changing to prefix tree